lib/elixir/lib/list.ex

```diff
@@ -897,50 +929,68 @@ defmodule List do
     compact_reverse(rest, [{kind, [elem]} | acc])
   end
 
-  defp each_diagonal(diag, limit, _paths, next_paths) when diag > limit do
+  defp each_diagonal(diag, limit, _paths, next_paths, _diff_script) when diag > limit do
     {:next, :lists.reverse(next_paths)}
   end
 
-  defp each_diagonal(diag, limit, paths, next_paths) do
-    {path, rest} = proceed_path(diag, limit, paths)
+  defp each_diagonal(diag, limit, paths, next_paths, diff_script) do
+    {path, rest} = proceed_path(diag, limit, paths, diff_script)
 
     case follow_snake(path) do
-      {:cont, path} -> each_diagonal(diag + 2, limit, rest, [path | next_paths])
+      {:cont, path} -> each_diagonal(diag + 2, limit, rest, [path | next_paths], diff_script)
       {:done, edits} -> {:done, edits}
     end
   end
 
-  defp proceed_path(0, 0, [path]), do: {path, []}
+  defp proceed_path(0, 0, [path], _diff_script), do: {path, []}
 
-  defp proceed_path(diag, limit, [path | _] = paths) when diag == -limit do
-    {move_down(path), paths}
+  defp proceed_path(diag, limit, [path | _] = paths, diff_script) when diag == -limit do
+    {move_down(path, diff_script), paths}
   end
 
-  defp proceed_path(diag, limit, [path]) when diag == limit do
-    {move_right(path), []}
+  defp proceed_path(diag, limit, [path], diff_script) when diag == limit do
+    {move_right(path, diff_script), []}
   end
 
-  defp proceed_path(_diag, _limit, [path1, path2 | rest]) do
+  defp proceed_path(_diag, _limit, [path1, path2 | rest], diff_script) do
     if elem(path1, 0) > elem(path2, 0) do
-      {move_right(path1), [path2 | rest]}
+      {move_right(path1, diff_script), [path2 | rest]}
     else
-      {move_down(path2), [path2 | rest]}
+      {move_down(path2, diff_script), [path2 | rest]}
     end
   end
 
-  defp move_right({y, list1, [elem | rest], edits}) do
+  defp move_right({y, [elem1 | rest1] = list1, [elem2 | rest2], edits}, diff_script)
+       when diff_script != nil do
+    if diff = diff_script.(elem1, elem2) do
+      {y + 1, rest1, rest2, [{:diff, diff} | edits]}
+    else
+      {y, list1, rest2, [{:ins, elem2} | edits]}
+    end
+  end
+
+  defp move_right({y, list1, [elem | rest], edits}, _diff_script) do
     {y, list1, rest, [{:ins, elem} | edits]}
   end
 
-  defp move_right({y, list1, [], edits}) do
+  defp move_right({y, list1, [], edits}, _diff_script) do
     {y, list1, [], edits}
   end
 
-  defp move_down({y, [elem | rest], list2, edits}) do
+  defp move_right({y, [elem1 | rest1], [elem2 | rest2] = list2, edits}, diff_script)
```

Comment:
This should be `move_down`.

Reply:
Also `move_right` (L#963) does `y + 1` in both branches, while `move_down` only with inner diff.

Reply:
> This should be move_down.

Thanks. Surprised the test suite didn't catch it. I guess it makes sense though since we always use the move_right branch until there are lefovers and the leftovers go to other branches.

> Also move_right (L#963) does y + 1 in both branches, while move_down only with inner diff.

`move_right` only moves y+1 when moving both lists.